redirector/go/main.go

```diff
@@ -0,0 +1,266 @@
+	DaemonVar:    "MOUNT_KBFUSE_DAEMON_PATH",
+}
+
+const (
+	mountpointTimeout = 5 * time.Second
+)
+
+type symlink struct {
+	link string
+}
+
+func (s symlink) Attr(ctx context.Context, a *fuse.Attr) (err error) {
+	a.Mode = os.ModeSymlink | a.Mode | 0555
+	a.Valid = 0
+	return nil
+}
+
+func (s symlink) Readlink(ctx context.Context, req *fuse.ReadlinkRequest) (
+	link string, err error) {
+	return s.link, nil
+}
+
+type cacheEntry struct {
+	mountpoint string
+	time       time.Time
+}
+
+type root struct {
+	lock            sync.RWMutex
+	mountpointCache map[uint32]cacheEntry
+}
+
+func newRoot() *root {
+	return &root{
+		mountpointCache: make(map[uint32]cacheEntry),
+	}
+}
+
+func (r *root) Root() (fs.Node, error) {
+	return r, nil
+}
+
+func (r *root) Attr(ctx context.Context, attr *fuse.Attr) error {
+	attr.Mode = os.ModeDir | 0555
+	return nil
+}
+
+func (r *root) getCachedMountpoint(uid uint32) string {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	entry, ok := r.mountpointCache[uid]
+	if !ok {
+		return ""
+	}
+	now := time.Now()
+	if now.Sub(entry.time) > mountpointTimeout {
+		// Don't bother deleting the entry, since the caller should
+		// just overwrite it.
+		return ""
+	}
+	return entry.mountpoint
+}
+
+func (r *root) findKBFSMount(ctx context.Context) (
+	mountpoint string, err error) {
+	uid := ctx.Value(fs.CtxHeaderUIDKey).(uint32)
+	// Don't let the root see anything here; we don't want a symlink
+	// loop back to this mount.
+	if uid == 0 {
+		return "", fuse.ENOENT
+	}
+
+	mountpoint = r.getCachedMountpoint(uid)
+	if mountpoint != "" {
+		return mountpoint, nil
+	}
+
+	defer func() {
+		if err != nil {
+			return
+		}
+		// Cache the entry if we didn't hit an error.
+		r.lock.Lock()
+		defer r.lock.Unlock()
+		r.mountpointCache[uid] = cacheEntry{
+			mountpoint: mountpoint,
+			time:       time.Now(),
+		}
+	}()
+
+	u, err := user.LookupId(strconv.FormatUint(uint64(uid), 10))
+	if err != nil {
+		return "", err
+	}
+
+	vols, err := gomounts.GetMountedVolumes()
+	if err != nil {
+		return "", err
+	}
+	fuseType := "fuse"
+	if runtime.GOOS == "darwin" {
+		fuseType = "kbfuse"
+	}
+	var fuseMountPoints []string
+	for _, v := range vols {
+		if v.Type != fuseType {
+			continue
+		}
+		if v.Owner != u.Uid {
+			continue
+		}
+		fuseMountPoints = append(fuseMountPoints, v.Path)
+	}
+
+	if len(fuseMountPoints) == 0 {
+		return "", fuse.ENOENT
+	}
+	if len(fuseMountPoints) == 1 {
+		return fuseMountPoints[0], nil
+	}
+
+	// If there is more than one, pick the first one alphabetically
+	// that has "keybase" in the path.
+	sort.Strings(fuseMountPoints)
+	for _, mp := range fuseMountPoints {
+		// TODO: a better regexp that will rule out keybase.staging if
+		// we're in prod mode, etc.
+		if !strings.Contains(mp, "keybase") {
+			continue
+		}
+		// Double-check that this is a real KBFS mount.
+		if _, err := os.Stat(filepath.Join(mp, ".kbfs_error")); err != nil {
```

Comment:
Curious why not `.kbfs_status`? Is this one cheaper?

Reply:
Yeah much cheaper, but anyway, I removed this.